Extra Jump API

absoluteAquarian opened the pull request:
### What is the new feature?
This pull request adds an actual API for adding modded extra jumps.  
The changes and additions that this API implements are as follows:
- `ExtraJumpState`
  - Stores the state of an extra jump for a player
  - Contains three properties:
    - `bool Enabled { get; }`
      - Set by `ExtraJumpState.Enable()`, cleared by `ExtraJumpState.Disable()`
    - `Available { get; set; }`
      - Indicates if an extra jump is ready to use (`true`) or has already been consumed (`false`). Jumps are consumed when they start
    - `Active { get; }`
      - Indicates whether the extra jump is currently being performed
    - `void Enable()`
      - Sets the extra jump as available to use, typically in equipment/accessory logic
      - `Disable()` overrides `Enable()` regardless of call order
    - `void Disable()`
      - Forcibly prevents using the extra jump for the current game tick. Consumes it if it was `Available`, and stops it if it was `Active`
- `ExtraJump`
  - A singleton `ModType` containing hooks related to performing visual effects for the jump and manipulating the player
  - `ExtraJump.Position GetDefaultPosition()`
    - Used to place the extra jump between **vanilla** extra jumps
    - If a non-vanilla extra jump is not returned here, an exception is thrown
    - There are a few predefined positions for modder convenience: `BeforeMountJumps`, `MountJumpPosition`, `BeforeBottleJumps` and `AfterBottleJumps`
  - `IEnumerable<ExtraJump.Position> GetModdedConstraints()`
    - Used to place the extra jump before or after other **modded** extra jumps that were sorted to the same "location" from `GetDefaultPosition()`
  - `void Visuals(Player player)`
    - Used to create effects while the extra jump is in progress, e.g. the dusts from the Blizzard in a Bottle's extra jump
  - `float GetDurationMultiplier(Player)`
    - Used to set the duration – or "strength" – of the extra jump
  - `void OnStarted(Player)`
    - Executes at the start of the extra jump
  - `void OnEnded(Player)`
    - Executes at the end of the extra jump, either from its duration expiring or when it is forcibly cancelled
  - `void UpdateHorizontalSpeeds(Player)`
    - Runs after `ModPlayer.PostUpdateMiscEffects()`, but before `ModPlayer.PostUpdateRunSpeeds()`
    - Used to adjust the horizontal speed variables (typically `runAcceleration` and `maxRunSpeed`) while the extra jump is in progress
  - `void OnRefreshed(Player)`
    - Executes whenever the extra jump is refreshed, typically when landing or being reeled in by a grappling hook
- `ExtraJumpLoader`
  - The corresponding loader for `ExtraJump`
- Changes to `Terraria.Player`:
  - All fields related to vanilla extra jumps have been integrated into a new `private ExtraJumpState[] extraJumps` field
    - Can be accessed via the `public Span<ExtraJumpState> ExtraJumps` property
    - Accessing the underlying array is similar to how `DamageClass` works:
      - `public ref ExtraJumpState GetJumpState<T>(T baseInstance) where T : ExtraJump`
      - `public ref ExtraJumpState GetJumpState<T>() where T : ExtraJump`
  - Flipper swimming has been reimplemented as an "extra jump" in this new API
    - The `accFlipper` field has been turned into a property which references `GetJumpState(ExtraJump.Flipper)`
  - The `sandStorm` field was turned into a property which indirectly depends on `GetJumpState(ExtraJump.SandstormInABottle).Active`
    - The reason for this change was due to [a gist](https://gist.github.com/absoluteAquarian/8a76bba28848ea5918bda5a26acf2f9a) documenting its purpose, which was essentially the same as whether the _Sandstorm in a Bottle_ jump is currently being performed (aka `Active`)
  - New methods/fields for facilitating mass checking and manipulation of extra jumps:
    - `bool AnyExtraJumpUsable()`
      - Returns `true` if any extra jump's `Available` property returns `true`
      - Setting `blockExtraJumps` (see below) to `true` will force this method to return `false`
    - `void ConsumeAllExtraJumps()`
      - Sets the `Available` property in all extra jumps to `false`
    - `bool blockExtraJumps`
      - Prevents extra jumps from being performed when set to `true` but does not consume them, or cancel an in progress jump
    - `void StopExtraJumpInProgress()`
      - Stops the current extra jump if one is active
  - `void RefreshAllExtraJumps()`
    - Previously was a `private` method named `RefreshDoubleJumps()`
    - Sets the `Available` property to `true` in any extra jumps that are `Enabled`
    - **NOTE:** disabled jumps via `ExtraJumpState.Disable()` cannot be refreshed
- Several new hooks to `ModPlayer` which are invoked in tandem with the corresponding hooks in `ExtraJump`

### Why should this be part of tModLoader?
Currently, adding an extra jump takes at minimum 3 IL edits and 3 method detours in order to properly implement, and doing so can easily clash with other mods.

### Are there alternative designs?
None that I can think of.

### Sample usage for the new feature
Indicating that a `ExtraJump` is available for use in `ModItem::UpdateAccessory()`:
```cs
public override void UpdateAccessory(Player player, bool hideVisual) {
	player.GetJumpState<SimpleExtraJump>().Enable();
}
```
Forcing the _Sandstorm in a Bottle_ extra jump to be disabled in `ModPlayer::PostUpdateMiscEffects()`:
```cs
public override void PostUpdateMiscEffects() {
	player.GetJumpState(ExtraJump.SandstormInABottle).Disable();
}
```
Preventing extra jumps while the player has a debuff
```cs
public virtual void Update(Player player, ref int buffIndex) {
	player.blockExtraJumps = true;
}
```

### ExampleMod updates
ExampleMod now includes the following:
- `Content/Items/Accessories/ExampleExtraJumpAccessory.cs`
  - showcases a basic extra jump
- `Content/Items/Accessories/ExampleMultiExtraJumpAccessory.cs`
  - showcases an advanced jump using a `ModPlayer` to track how many jumps the extra jump has remaining
- `Common/Players/ExampleExtraJumpModificationPlayer.cs`
  - showcases an example of modifying an extra jump

### Additional Notes
If a modder wants to modify the `ExtraJump.UpdateHorizonalSpeeds()` changes from an ExtraJump, they should use the following in ModPlayer:
```cs
public override void PostUpdateRunSpeeds() {
		if (Player.GetJumpState<TheJump>().Active && !Player.Mount.Active) {
				// Revert speed changes here
				// ... then apply custom changes
		}
		// Additional logic goes here
}
```

### Bug Fixes
The extra jump for the Basilisk mount was meant to have the highest priority, but due to a small code error, it only has priority over the Cloud in a Bottle extra jump _**and**_ it ends up being consumed, even if it wasn't actually used.

This PR fixes the bug by restoring the mount's proper priority order (basilisk > goat) and preventing it from being always consumed.

absoluteAquarian committed:
`ModExtraJump`, its related types and many changes to `Player`

absoluteAquarian committed:
remove unnecessary `using` statements

absoluteAquarian committed:
properly match vanilla logic when performing extra jumps

absoluteAquarian committed:
added TODO message

absoluteAquarian committed:
added `ModExtraJump.IgnoresSwimmingChecks` and the hardcoded consumption for the Basilisk mount's extra jump

absoluteAquarian committed:
comment typo

absoluteAquarian committed:
added `ModExtraJump.OnJumpEnded`

absoluteAquarian committed:
remove TODO comment

absoluteAquarian committed:
implemented extra jump order priority, needs testing

absoluteAquarian committed:
`sealed` the vanilla extra jump classes

absoluteAquarian committed:
forgot to remove erroneous override in `GoatMountJump`

absoluteAquarian committed:
bug fixes and other changes

- fixed some bugs which caused extra jumps to straight-up not work at all
-  ensure that `PlayerMovementAccsCache` respects the existance of modded jumps
- make rockets only triggerable when all extra jumps have been consumed (previously only checked the Cloud in a Bottle flag)

absoluteAquarian committed:
added `ModExtraJump.OnJumpRefreshed` and changed some of the methods in `Player`

absoluteAquarian committed:
consistent newline spacing from mass comment patching

absoluteAquarian committed:
more consistent spacing

absoluteAquarian committed:
added `AfterParent` and `BeforeParent` for positioning, optimized code responsible for ordering extra jumps

absoluteAquarian committed:
change TopoSort to `int` sorting since that has a better `GetHashCode()` implementation

absoluteAquarian committed:
rename `BeforeParent` and `AfterParent` to `Before` and `After`

absoluteAquarian committed:
moved jump priority order to mod loading rather than during runtime

- renamed `GetOrder` to `GetDefaultPosition`
- `GetDefaultPosition` no longer take a `Player` argument
- added `GetModdedConstraints`, which runs after the overall order in regard to vanilla extra jumps has been fully processed
- made all methods in `ExtraJumpLoader` that use `ModExtraJump` hooks run in the order of `OrderedJumps` rather than `ExtraJumps`

absoluteAquarian committed:
added missing summaries and logic

- added `ExtraJumpLoader.ResetActiveFlags()` for resetting all extra jumps in `Player.ResetEffects()`
- added `ExampleExtraJumpAccessory`
- added `ExampleMultiExtraJumpAccessory`
- expanded the XML summary for `ModExtraJump.ModifyPlayerHorizontalSpeeds()`
- renamed `ExtraJumpLoader.HandleJumpVisuals()` to `JumpVisuals()`
- added `ModExtraJump.PreJumpVisuals()`
- made `ModExtraJump.OnJumpRefreshed()` always run without regard to the current state of `JumpAvailable`
- fixed the `ExtraJumpLoader.OrderedJumps` sorting

absoluteAquarian committed:
previous commit did not have most of the changes

absoluteAquarian committed:
added `GlobalExtraJump`

absoluteAquarian committed:
small changes

absoluteAquarian committed:
added `SandstormJumpModification`

absoluteAquarian committed:
implemented suggestions

- renamed `ModExtraJump` to `ExtraJump`
- added an example of using `ExtraJump.GetModdedConstraints()` to `SimpleExtraJump`
- updated the XML summary for `GetModdedConstraints()`
- properly support `return null;` for `GetModdedConstraints()`

absoluteAquarian committed:
change `yield break;` to `return null;`

absoluteAquarian committed:
various API changes

- removed `GlobalExtraJump` and moved all of its hooks into `ModPlayer`
- removed `ExtraJump.IgnoreSwimmingChecks` since that was never vanilla behavior and i just misread the code
- renamed all `ExtraJump` methods to not include "Jump" in the name since that's redundant
- added `ExtraJump.OnCleared()`

absoluteAquarian committed:
added sprites for the example accessories

absoluteAquarian committed:
removed unnecessary `Player player` parameter from new `ModPlayer` hooks

absoluteAquarian committed:
forgot to remove a param summary

absoluteAquarian committed:
added `ModPlayer.ExtraJumpVisuals()`

Chicken-Bones commented on ExampleMod/Content/Items/Accessories/ExampleExtraJumpAccessory.cs:
```diff
@@ -0,0 +1,123 @@
+
+			Vector2 spawnPos = new Vector2(player.position.X, player.position.Y + offsetY);
+
+			for (int i = 0; i < 2; i++) {
+				Dust dust = Dust.NewDustDirect(spawnPos, player.width, 12, DustID.Snow, player.velocity.X * 0.3f, player.velocity.Y * 0.3f, newColor: Color.Gray);
+				dust.velocity *= 0.1f;
+				if (i == 0)
+					dust.velocity += player.velocity * 0.03f;
+				else
+					dust.velocity -= player.velocity * 0.03f;
+
+				dust.velocity -= player.velocity * 0.1f;
+				dust.noGravity = true;
+				dust.noLight = true;
+			}
+
+			for (int i = 0; i < 3; i++) {
+				Dust dust = Dust.NewDustDirect(spawnPos, player.width, 12, DustID.Snow, player.velocity.X * 0.3f, player.velocity.Y * 0.3f, newColor: Color.Gray);
+				dust.fadeIn = 1.5f;
+				dust.velocity *= 0.6f;
+				dust.velocity += player.velocity * 0.8f;
+				dust.noGravity = true;
+				dust.noLight = true;
+			}
+
+			for (int i = 0; i < 3; i++) {
+				Dust dust = Dust.NewDustDirect(spawnPos, player.width, 12, DustID.Snow, player.velocity.X * 0.3f, player.velocity.Y * 0.3f, newColor: Color.Gray);
+				dust.fadeIn = 1.5f;
+				dust.velocity *= 0.6f;
+				dust.velocity -= player.velocity * 0.8f;
+				dust.noGravity = true;
+				dust.noLight = true;
+			}
```
don't like playing spot the difference, copy paste bad, please extract to a method with relevant parameters (`velocityMultiplier, playerVelocityMultiplier`)

absoluteAquarian replied:
noted, i'll abstract it to a method

Chicken-Bones commented on patches/tModLoader/Terraria/Player.TML.cs:
```diff
@@ -594,4 +594,40 @@ public void AddImmuneTime(int cooldownCounterId, int immuneTime)
+	// Extra jumps
+	private ExtraJumpData[] extraJumps = new ExtraJumpData[ExtraJumpLoader.ExtraJumpCount];
+
+	public ref ExtraJumpData GetExtraJump<T>(T baseInstance) where T : ExtraJump => ref extraJumps[baseInstance.Type];
+
+	public ref ExtraJumpData GetExtraJump<T>() where T : ExtraJump => ref GetExtraJump(ModContent.GetInstance<T>());
```
Should we call the class `State` instead of `Data`? 

Not a big fan of the method returning a different type to what the name implies. `GetJumpState` or `GetJumpData` better perhaps?

absoluteAquarian replied:
renamed to `ExtraJumpState` and `GetJumpState()`

Chicken-Bones commented on patches/tModLoader/Terraria/Player.TML.cs:
```diff
@@ -594,4 +594,40 @@ public void AddImmuneTime(int cooldownCounterId, int immuneTime)
+	{
+		for (int i = 0; i < extraJumps.Length; i++)
+		{
+			ExtraJumpData data = extraJumps[i];
```
`foreach (var data in extraJumps)`

Use `ref var` if you need to modify the value in-loop

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/VanillaExtraJumps.cs:
```diff
@@ -0,0 +1,401 @@
+		Main.gore[num25].velocity.Y = Main.gore[num25].velocity.Y * 0.1f - player.velocity.Y * 0.05f;
+	}
+
+	public override bool PreVisuals(Player player) => (player.canJumpAgain_Sandstorm || !player.hasJumpOption_Sandstorm) && base.PreVisuals(player);
```
comment, what does this do?

absoluteAquarian replied:
allows modders to override/expand the requirements for making the jump visuals appear

Chicken-Bones replied:
More specifically, why does sandstorm have this special case?

absoluteAquarian replied:
this is from `CloudInABottleJump`, and it's present because the vanilla code in `Player::DoubleJumpVisuals()` also had that for some reason

absoluteAquarian replied:
removed due to being obscure and not having any meaningful impact on gameplay other than blocking the Cloud jump's dust particles for no reason

Chicken-Bones commented on patches/tModLoader/Terraria/DataStructures/ExtraJumpData.cs:
```diff
@@ -0,0 +1,20 @@
+	/// <summary>
+	/// Whether the extra jump has been consumed
+	/// </summary>
+	public bool JumpAvailable;
```
Names `Active` and `JumpAvailable` are unclear. Documentation here actually implies inverse of correct use.
Consider 
`Active` -> `Equipped` or `Enabled`
`JumpAvailable` -> `Consumed` (and inverting the associated logic)

Documentation also needs to be clear about _when_ the flags are set. Does the `Consumed` get set immediately on pressing space?

What happens if a jump is unequipped while `PerformingJump`?

absoluteAquarian replied:
renamed `Active` to `Enabled`

`JumpAvailable` is mimicking the vanilla behaviour, which uses this flag for saying whether the jump can be used... aka it's "available" for consumption.  inverting the logic would be a headache for not much gain

`JumpAvailable` gets cleared immediately when the extra jumps starts (aside from contexts where all jumps are cleared)

unequipping an extra jump while `PerformingJump` is true just makes the visuals not appear.  re-equipping the jump while the "jump duration" is still in effect just makes the visuals appear again (this is the vanilla behaviour)

Chicken-Bones replied:
Needs to be included in documentation here. `Whether the extra jump has been consumed` implies that it's true if consumed. I don't think people will actually make that mistake, because the variable name is clear, but consider 
```
True if the jump has not yet been used. Will be set to false when a jump starts.
```
also:
```
Remember to also check Enabled
```

absoluteAquarian replied:
noted.  i'll change the summary

Chicken-Bones commented on patches/tModLoader/Terraria/DataStructures/ExtraJumpData.cs:
```diff
@@ -0,0 +1,20 @@
+	/// <summary>
+	/// Whether any effects (e.g. spawning dusts) should be performed
+	/// </summary>
+	public bool PerformingJump;
```
If `Active` is the only one meant to be set by modders, the rest should be `{ get; internal set; }`. Consider making them all properties for uniformity/future change proofing.

absoluteAquarian replied:
all of the corresponding vanilla flags are `public`, so i don't see why `JumpAvailable` and `PerformingJump` should be given limited set accessibility

Chicken-Bones replied:
Vanilla doesn't set good precedents for an API. The point of limiting set accessibility is to guide modders away from bugs. If these flags are adequately controlled by tML, modders don't have to think about why or when to change them. They also don't have to think about what may happen if they get changed at a weird time.

absoluteAquarian replied:
i can see why `PerformingJump` should be limited, but imo modders should still be able to modify `JumpAvailable`

Chicken-Bones replied:
I recall one of the uses was marking a jump as still having uses left. Just need to note in the docs _why_ you would consider setting it to true or false, and what happens if you set it at a specific time (with respect to the other flags and logic)

absoluteAquarian replied:
added documentation

Chicken-Bones commented on patches/tModLoader/Terraria/DataStructures/ExtraJumpData.cs:
```diff
@@ -0,0 +1,20 @@
+public struct ExtraJumpData
+{
+	/// <summary>
+	/// Whether the extra jump can be used. This field is what should be set by equipment
```
Need to document _when_ it should be set. The fact it's automatically cleared in `ResetStats`. If mods are only meant to enable it, then make a method `Enable()`. 

Don't want mods fighting over enable/disable order.

Where are mods meant to disable jumps. I see two use cases:
- Conflict (a modder doesn't want a certain jump to work when another accessory is equipped or action is being performed)
- Temporary lockout to all jumps (debuff, or a state which prevents jumping entirely like being in a tractor beam or something). If the lockout ends while the player still has air-time, they should be able to continue using jumps they haven't yet used.

Ideally need examples of the above/documentation on how to correctly acheive

absoluteAquarian replied:
documentation added.

considering adding a `Disable()` method which uses a hidden backing field to make the "value" of `Enabled` (formerly `Active`) appear as `false` even when it's set to `true` by making it a property

this should remedy any conflicts for disabling jumps (setting `Active` to `false` can still disable it, calling `Disable()` forcibly disables it)

Chicken-Bones replied:
Yeah, enabled overrides are good, we use this design with `SetCrit()` and `DisableCrit()` in hit modifiers for example, providing a clear override precedence.

You've explained to me that setting `Enabled = false` doesn't clear `JumpAvailable`, so that when the thing disabling the jump expires, the jumps can continue where they left off. This is good, but should be clear to a modder (from documentation on `JumpState`)

I would also recommend adding a variable to `Player` which can disable all double jumps (and maybe another one for flight), as a convenience so a modder doesn't have to loop over all jumps every frame to disable them.

We could use such a flag to implement the flipper swimming jump lockout for example.

absoluteAquarian replied:
decided to add a `ExtraJumpState::Disable()` method which works as a forced disabling of the extra jump (prevents `PerformJump` from executing, does not affect `Enabled`) as well as `Player::blockExtraJumps` for easily disabling all extra jumps (it's checked alongside the flipper check)

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ExtraJumpLoader.cs:
```diff
@@ -0,0 +1,224 @@
+	public static void ModifyPlayerHorizontalSpeeds(Player player)
+	{
+		foreach (ExtraJump moddedExtraJump in orderedJumps) {
+			// Special case: Sandstorm in a Bottle uses a separate flag
```
anything we can do about this?

absoluteAquarian replied:
frustratingly, no
for whatever reason, `sandStorm` has separate and unique checks and assignments apart from `isPerformingJump_Sandstorm`

Chicken-Bones replied:
Can you do some research, and make a write-up? What's special about sandstorm, aside from perhaps janky code. What does it do differently to blizzard/tsunami? How would a mod replicate those effects if they wanted to.

absoluteAquarian replied:
sure, i'll add another comment here — or a gist — with my findings (and a mention in the discord thread) once i get around to that

absoluteAquarian replied:
the gist in question: https://gist.github.com/absoluteAquarian/8a76bba28848ea5918bda5a26acf2f9a

absoluteAquarian replied:
resolved due to making `Player.sandStorm` rely on `Player.isPerformingJump_Sandstorm`

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ExtraJumpLoader.cs:
```diff
@@ -0,0 +1,224 @@
+
+	public static void ProcessJumps(Player player, bool flipperOrSlimeMountSwimming)
+	{
+		if (!flipperOrSlimeMountSwimming) {
```
invert if, early return. 

Does this condition belong in here, or outside? Is this an example of one of the the 'all jump lockout' reasons I mentioned above?

absoluteAquarian replied:
leftover from when there was a property to ignore that parameter; there's no need for the parameter anymore

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ExtraJumpLoader.cs:
```diff
@@ -0,0 +1,224 @@
+	{
+		foreach (ExtraJump jump in ExtraJumps) {
+			ref ExtraJumpData data = ref player.GetExtraJump(jump);
+			if (!data.Active) {
```
invert if, early return

absoluteAquarian replied:
i don't think that you understand what this method is doing.
the vanilla logic had several blocks of
```cs
if (!hasJumpOption_X)
    canJumpAgain_X = false;
```
which this method just abstracts to a loop

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ExtraJumpLoader.cs:
```diff
@@ -0,0 +1,224 @@
+		}
+	}
+
+	public static void RefreshExtraJumps(Player player)
```
what if a jump gets equipped mid flight? Should it become available? Probably not I suppose.

absoluteAquarian replied:
vanilla has one case where they can be publicly refreshed (`Player::RefreshMovementAbilities()`).  every other case is either from standing on the ground or jumping out of a grapple.
should that Player method be made `internal` along with this ExtraJumpLoader method?

Chicken-Bones replied:
The `Active` check here means that if something is suppressing a double-jump when the player leaves the ground, but then expires, the jump will not be considered available :(

absoluteAquarian replied:
alas, this is a mimicry of vanilla behavior

granted, vanilla never had anything except for a hardcoded flipper check for disabling jumps.  this could probably be remedied by the `Disable()` method stuff mentioned in another conversation

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ExtraJump.cs:
```diff
@@ -0,0 +1,157 @@
+	/// By default, this hook returns whether the player is moving upwards with respect to <see cref="Player.gravDir"/>
+	/// </summary>
+	/// <param name="player">The player performing the jump</param>
+	public virtual bool PreVisuals(Player player) {
```
Consider `CanShowVisuals`?

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ExtraJump.cs:
```diff
@@ -0,0 +1,157 @@
+	/// This hook runs before the <see cref="ExtraJumpData.JumpAvailable"/> flag for this extra jump is set to <see langword="false"/> in <see cref="Player.Update(int)"/> due to the jump being unavailable or when calling <see cref="Player.ClearAllExtraJumps"/> (vanilla calls it when a mount that blocks jumps is active).
+	/// </summary>
+	/// <param name="player">The player instance</param>
+	public virtual void OnCleared(Player player) { }
```
Please provide a reason to use the hook in the documentation

absoluteAquarian replied:
removed due to not being able to think of a good use case

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ExtraJump.cs:
```diff
@@ -0,0 +1,157 @@
+	/// </para>
+	/// </summary>
+	/// <param name="player">The player performing the jump</param>
+	public virtual void ModifyHorizontalSpeeds(Player player) { }
```
Prefer a more general name, perhaps `Update`

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/ExtraJump.cs:
```diff
@@ -0,0 +1,157 @@
+	/// </summary>
+	/// <param name="player">The player performing the jump</param>
+	/// <returns>A modifier to the player's jump height, which when combined effectively acts as the duration for the extra jump</returns>
+	public abstract float GetDuration(Player player);
```
`GetJumpHeightMultiplier`?

absoluteAquarian replied:
most users may be able to understand it setting a "duration" easier, but i'm willing to change it if saying it's a "multiplier" would be better

Chicken-Bones replied:
duration is fine, but `Multiplier` needs to be in the name. Since `1.0` isn't a duration

absoluteAquarian replied:
good point

Chicken-Bones commented on patches/tModLoader/Terraria/Player.TML.cs:
```diff
@@ -594,4 +594,40 @@ public void AddImmuneTime(int cooldownCounterId, int immuneTime)
+		return false;
+	}
+
+	// An alternative more in line with the vanilla behavior
```
Don't like the need for this, hard to reason about.

absoluteAquarian replied:
this is just to keep the vanilla logic as it was in order to future proof against any inconsistencies (this is also why other quirks were left as-is)

Chicken-Bones replied:
To keep this maintainable, we need to know why you would want to use this method rather than the one above. I understand "that's what vanilla had" but we're expanding on it here, so we need to be able to fully understand the system (vanilla isn't known for easy to understand code). 

I would expect that checking if jumps are available without an active check would result in potential bugs. Maybe the active check isn't _needed_ in the vanilla calls but it wouldn't hurt? I haven't looked at the logic deeply, but it seems spooky to me.

absoluteAquarian replied:
in the vanilla logic, it's specifically checking if all extra jumps have been consumed (`JumpAvailable` would be `false` for all jumps).  furthermore, an extra jump is "consumed" by default

so the `Enabled` check would likely just end up with the same result?  after all, a disabled jump is always "consumed" and enabled jumps have to be used to be consumed... so checking `Enabled` would be pointless in a pure vanilla context

however, if modders can set `JumpAvailable` to true, then this would throw a wrench in the plan when the extra jump's `Enabled` flag is cleared

absoluteAquarian replied:
resolved by removing the `NoActiveCheck` method in favor of making vanilla check `Enabled` as well via the `public` method

absoluteAquarian committed:
implemented suggestions

- renamed `ExtraJumpData` to `ExtraJumpState`
- renamed `Player::GetExtraJump()` to `Player::GetJumpState()`
- renamed various methods
- removed unnecessary hooks
- abstracted the code in `ExampleExtraJumpAccessory.cs`

absoluteAquarian committed:
updated XML summaries for the fields in `ExtraJumpState`

absoluteAquarian committed:
rename `GetDuration` to `GetDurationMultiplier`

absoluteAquarian committed:
added logic for disabling extra jumps

- added `ExtraJumpState::Disable()` for forcing an extra jump to be disabled.  a disabled jump cannot be re-enabled for the current game tick
- added `Player::blockExtraJumps` for easily preventing all extra jumps from being usable
- added or expanded XML summaries in various locations

absoluteAquarian committed:
remove obscure `JumpVisuals` constraint for the Cloud jump

absoluteAquarian committed:
added documentation

absoluteAquarian committed:
updated documentation

absoluteAquarian committed:
various API changes

- made `ExtraJumpState.PerformingJump` publicly get-only
- added/updated various XML summaries
- renamed several methods in `ExtraJumpLoader`
- made `Player.CancelAllExtraJumpVisuals()` hidden again
- added `Player.StopExtraJumpInProgress()`
- modified `Player.sandStorm` so that the Sandstorm in a Bottle doesn't need a special case for handling movement

absoluteAquarian committed:
forgot to remove comment

absoluteAquarian committed:
add tModPorter logic for replacing the player fields with GetJumpState usage

absoluteAquarian committed:
updated XML summaries and removed `Player.AnyExtraJumpAvailableNoActiveCheck()`

absoluteAquarian committed:
updated XML summary

absoluteAquarian committed:
updated XML summaries

absoluteAquarian committed:
updated XML summary

absoluteAquarian committed:
added `ExtraJumpState.Enable()` in order to prevent `Enabled` from being "set" to `false` outside of tML scope

absoluteAquarian committed:
reorganized some logic and added state validity checks to `ExtraJumpState`

absoluteAquarian committed:
various adjustments to how the properties in `ExtraJumpState` works

absoluteAquarian committed:
property renames, logic tweaks, updated ExampleMod, updated tModPorter

absoluteAquarian committed:
method renames

absoluteAquarian committed:
logic tweaks

absoluteAquarian committed:
`Before/After.Parent` -> `Before/After.Target`

absoluteAquarian committed:
cleaned up jump sorting logic

absoluteAquarian committed:
cleaned up sorting logic, added `ExtraJumpLoader.OrderedJumps`

absoluteAquarian committed:
make disabled jumps be automatically consumed and not refreshable

absoluteAquarian committed:
fix erroneous upper bound from previous iteration of code

absoluteAquarian committed:
added missing localization for the two new accessories

absoluteAquarian committed:
update summary for `ExtraJumpState.Disable()`

absoluteAquarian committed:
update summary again

absoluteAquarian committed:
bug fixes and localization updates

absoluteAquarian committed:
updated XML summary

Chicken-Bones committed:
Fix tModPorter tests

absoluteAquarian committed:
various internal logic changes, removed unnecessary properties in `Player`

absoluteAquarian committed:
fix vanilla bugs rather than leaving them in

absoluteAquarian committed:
added `CanStart` hook for conditional jumps

Chicken-Bones committed:
ExtraJumpState docs improvement

absoluteAquarian committed:
added better predefined `ExtraJump.Position` values

Chicken-Bones committed:
Visuals -> ShowVisuals

Chicken-Bones committed:
ExtraJump doc improvements

Chicken-Bones committed:
Restore accidentally deleted CanStart

Chicken-Bones committed:
Change some loader members to private

Chicken-Bones committed:
Restore missing hasJumpOption_WallOfFleshGoat

absoluteAquarian committed:
make `AnyExtraJumpAvailable()` check CanStart

absoluteAquarian committed:
updated XML summary

Chicken-Bones committed:
Better ConsumeAllExtraJumps doc

absoluteAquarian committed:
updated XML summary

absoluteAquarian committed:
reimplement flipper swimming as an "extra jump"

absoluteAquarian committed:
added tModPorter case for `accFlipper` changes

absoluteAquarian committed:
updated XML summaries

absoluteAquarian committed:
updated XML summary

absoluteAquarian committed:
made `accFlipper` public again, setting to `false` now calls `Disable()` for `FlipperJump`

absoluteAquarian committed:
updated tModPorter suggestion

absoluteAquarian committed:
rename `Player.RefreshDoubleJumps()` to `RefreshAllExtraJumps`

absoluteAquarian committed:
RefreshAllExtraJump -> RefreshExtraJumps, also updated the XML summary for it

absoluteAquarian committed:
`Player.AnyExtraJumpAvailable()` -> `AnyExtraJumpUsable()`
